neurodamus/utils/memory.py

```diff
@@ -568,3 +578,90 @@ def validate_inputs_distribute(self, num_ranks, batch_size):
         #       Also Counter() will discard 0-count entries
         # syn_count_metypes = set(self.metype_cell_syn_average)
         # assert all_metypes <= syn_count_metypes, all_metypes - syn_count_metypes
+
+    def check_all_buckets_have_gids(self, bucket_allocation, num_ranks, cycles):
+        """
+        Checks if all possible buckets determined by num_ranks and cycles have at least one GID
+        assigned.
+
+        Args:
+            bucket_allocation (dict): The allocation dictionary containing the assignment of GIDs
+                                      to ranks and cycles.
+            num_ranks (int): The number of ranks.
+            cycles (int): The number of cycles.
+
+        Returns:
+            bool: True if all buckets have at least one GID assigned, False otherwise.
+        """
+        for pop, rank_allocation in bucket_allocation.items():
+            for rank_id in range(num_ranks):
+                for cycle_id in range(cycles):
+                    if not rank_allocation.get((rank_id, cycle_id)):
+                        logging.warning(f"Bucket ({rank_id}, {cycle_id}) in population '{pop}' "
+                                        f"has no GIDs assigned.")
+                        return False
+        return True
+
+    @run_only_rank0
+    def distribute_cells_with_validation(self,
+                                         num_ranks,
+                                         cycles=None,
+                                         metype_file=None
+                                         ) -> Tuple[dict, dict, dict]:
+        """
+        Wrapper function to distribute cells with ever smaller assignment batches until
+        all buckets have at least one GID assigned (or fail if no valid distribution is found).
+
+        Args:
+            num_ranks (int): The number of ranks.
+            cycles (int): The number of cycles to distribute cells over.
+            metype_file (str): The path to a JSON file containing memory usage for each METype.
+
+        Returns:
+            Tuple[dict, dict, dict]: Returns the same as distribute_cells once a valid distribution
+                                     is found.
+        """
```

Comment:
Maybe some logging would be nice at the beginning of the function?
Something like
logging.info(f"Starting distribution with {num_ranks} ranks and {cycles} cycles.")